### ISSUE TITLE
addEventListener hook?

### ISSUE THREAD

**Issue by matthewjh** [angular/zone.js#142]
Is the `addEventListener` hook implemented? I couldn't get it to work at all.

**Comment by vicb** [comment]
What are you trying to do ?

What browser, what version of zone.js, do you have a code sample ?

**Comment by matthewjh** [comment]
```javascript
	    var myZone = zone.fork({
	        addEventListener: function() {
	            console.log('addEventListener hook: ', arguments);
	        }
	    });
	    myZone.run(function() {
	        window.addEventListener('click', function() {
	            console.log('click!');
	        });
	    });
```

When I run that snippet with the master dist copy of zone.js loaded in, `"click!"` is logged when you click in the window but `"addEventListener hook"` isn't. I searched the codebase for `zone.addEventListener` and it turned up nothing, hence my asking whether this feature had been implemented.

If it hasn't been implemented, I'm interested in submitting a PR for it, but I wanted to check first.

**Comment by btford** [comment]
I think you want:

```
        var myZone = zone.fork({
            afterTask: function() {
                console.log('addEventListener hook: ', arguments);
            }
        });
        myZone.run(function() {
            window.addEventListener('click', function() {
                console.log('click!');
            });
        });
```

Zone is a generic mechanism that fires on any async operation. Are you just trying to capture clicks? What's your specific use case?

**Comment by matthewjh** [comment]
Here, I have no specific use case. I'm just following the readme which explicitly says there's an addEventListener hook, which it says "allows you to intercept calls to EventTarget.addEventListener"

**Comment by matthewjh** [comment]
My use case is that of someone following the readme. :p

**Comment by vicb** [comment]
If you don't have any specific use case, may be the best would be to drop this feature (from the doc).

My concern is that `patchEventTargetMethods()` [is used in different ways](https://github.com/angular/zone.js/blob/master/lib/patch/event-target.js) according to what the browser supports.

@btford what do you think ?

**Comment by matthewjh** [comment]
A very general use case would be if you want to do some action before or after a task, or otherwise wrap the callback function, but only for event listeners. What if you wanted to report all effecting events in an app to an analytics service? What if you want custom error handling for event listeners? IMO it could be pretty useful to have. 

More generally, it would be cool if you could supply custom patches for `add/removeEventListener`, `setTimeout`, `alert`, etc., on a per-zone basis, which is what the readme seems to hint at with interceptors.

**Comment by vicb** [comment]
> What if ...

That's not something we need for now. Hence my proposal to remove it from the doc now. At some later point if we need it (or if someone works on the impl) we can restore the doc.

> More generally, it would be cool if you could supply custom patches for `add/removeEventListener`, `setTimeout`, `alert`, etc., on a per-zone basis

IIRC `setTimeout` (and the like) & `alert` are supported. only `add/removeEventListener` is not.

**Comment by matthewjh** [comment]
"or if someone works on the impl"

Well, like I said, I'd be happy to take a stab at it.

**Comment by vicb** [comment]
> Well, like I said, I'd be happy to take a stab at it.

Then you should :)

**Comment by btford** [comment]
> Well, like I said, I'd be happy to take a stab at it.

Yep! Go for it.

Until then, we should probably fix the readme.

**Comment by scamden** [comment]
yep just spent an hour or so scouring the code for this hook's impl till I found this issue. my use case is to fork a new zone in specific event callbacks in order to track user actions, so tracking the timestamps of all the resulting js execution of one user's click for example

**Comment by matthewjh** [comment]
That use case is very similar to my original use case. 

I'll work on this again later today.

**Comment by scamden** [comment]
Awesome!

**Comment by scamden** [comment]
@matthewjh let me know if i can help

**Comment by matthewjh** [comment]
Implemented by #171.
